prov/efa/src/rdm/efa_rdm_ep_utils.c

```diff
@@ -675,6 +675,281 @@ void efa_rdm_ep_post_handshake_or_queue(struct efa_rdm_ep *ep, struct efa_rdm_pe
+ *     local read-copy packet pool (rx_readcopy_pkt_pool).
+ *
+ * This function is called when the progress engine is called for
+ * the 1st time on this endpoint.
+ *
+ * @param ep[in,out]	endpoint
+ * @return		On success, return 0
+ * 			On failure, return a negative error code.
+ */
+int efa_rdm_ep_grow_rx_pools(struct efa_rdm_ep *ep)
+{
+	int err;
+
+	assert(ep->efa_rx_pkt_pool);
+	err = ofi_bufpool_grow(ep->efa_rx_pkt_pool);
+	if (err) {
+		EFA_WARN(FI_LOG_CQ,
+			"cannot allocate memory for EFA's RX packet pool. error: %s\n",
+			strerror(-err));
+		return err;
+	}
+
+	if (ep->rx_unexp_pkt_pool) {
+		assert(ep->rx_unexp_pkt_pool);
+		err = ofi_bufpool_grow(ep->rx_unexp_pkt_pool);
+		if (err) {
+			EFA_WARN(FI_LOG_CQ,
+				"cannot allocate memory for unexpected packet pool. error: %s\n",
+				strerror(-err));
+			return err;
+		}
+	}
+
+	if (ep->rx_ooo_pkt_pool) {
+		assert(ep->rx_ooo_pkt_pool);
+		err = ofi_bufpool_grow(ep->rx_ooo_pkt_pool);
+		if (err) {
+			EFA_WARN(FI_LOG_CQ,
+				"cannot allocate memory for out-of-order packet pool. error: %s\n",
+				strerror(-err));
+			return err;
+		}
+	}
+
+	if (ep->rx_readcopy_pkt_pool) {
+		err = ofi_bufpool_grow(ep->rx_readcopy_pkt_pool);
+		if (err) {
+			EFA_WARN(FI_LOG_CQ,
+				"cannot allocate and register memory for readcopy packet pool. error: %s\n",
+				strerror(-err));
+			return err;
+		}
+	}
+
+	if (ep->map_entry_pool) {
+		err = ofi_bufpool_grow(ep->map_entry_pool);
+		if (err) {
+			EFA_WARN(FI_LOG_CQ,
+				 "cannot allocate memory for map entry pool. error: %s\n",
+				 strerror(-err));
+		}
+
+		return err;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief post internal receive buffers for progress engine.
+ *
+ * It is more efficient to post multiple receive buffers
+ * to the device at once than to post each receive buffer
+ * individually.
+ *
+ * Therefore, after an internal receive buffer (a packet
+ * entry) was processed, it is not posted to the device
+ * right away.
+ *
+ * Instead, we increase counter
+ *      ep->efa_rx_pkts_to_post
+ * by one.
+ *
+ * Later, progress engine calls this function to
+ * bulk post internal receive buffers (according to
+ * the counter).
+ *
+ * This function also control number of internal
+ * buffers posted to the device in zero copy receive
+ * mode.
+ *
+ * param[in]	ep	endpoint
+ */
+void efa_rdm_ep_post_internal_rx_pkts(struct efa_rdm_ep *ep)
+{
+	int err;
+
+	if (ep->use_zcpy_rx) {
+		/*
+		 * In zero copy receive mode,
+		 *
+		 * If application did not post any receive buffer,
+		 * we post one internal buffer so endpoint can
+		 * receive control packets such as handshake.
+		 *
+		 * If buffers have posted to the device, we do NOT
+		 * repost internal buffers to maximize the chance
+		 * user buffer is used to receive data.
+		 */
+		if (ep->efa_rx_pkts_posted == 0 && ep->efa_rx_pkts_to_post == 0) {
+			ep->efa_rx_pkts_to_post = 1;
+		} else if (ep->efa_rx_pkts_posted > 0 && ep->efa_rx_pkts_to_post > 0){
+			ep->efa_rx_pkts_to_post = 0;
```

Comment:
Effectively the same logic; and a little clearer based on the comment above (unless `ep->efa_rx_pkts_to_post` can be negative):

```suggestion
		} else if (ep->efa_rx_pkts_posted > 0) {
			ep->efa_rx_pkts_to_post = 0;
```

Reply:
Good catch. Actually it is a bug and your suggested change is the correct fix. I already fixed that in my zero copy recv branch. Will update it here as well.